linera-service/src/linera.rs

```diff
@@ -1260,11 +1279,12 @@ impl Runnable for Job {
                 info!("Starting operation to open a new chain");
                 let time_start = Instant::now();
                 let ownership = ChainOwnership::single(new_public_key);
-                let (message_id, certificate) = chain_client.open_chain(ownership, balance).await?;
+                let result = chain_client.open_chain(ownership, balance).await;
+                context.update_and_save_wallet(&mut chain_client).await;
+                let (message_id, certificate) = result.context("failed to open chain")?;
                 let time_total = time_start.elapsed().as_micros();
                 info!("Operation confirmed after {} us", time_total);
                 debug!("{:?}", certificate);
-                context.update_wallet_from_client(&mut chain_client).await;
                 let id = ChainId::child(message_id);
                 let timestamp = match certificate.value() {
                     CertificateValue::ConfirmedBlock {
```

Comment:
Yes, but we have to update and save even if `open_chain` returned an error, because it might have changed the pending block. Same reason in all the places below.

I wonder whether it would be cleaner to change the `ChainClient` API and make all the commands like `open_chain`, `transfer`, etc. only set the pending block, without trying to send it? Then the code here would look more like:

```rust
// Set the pending block to include an `OpenChain` operation.
chain_client.open_chain(ownership, balance).await?;
// Make sure the pending block gets saved.
context.update_and_save_wallet(&mut chain_client).await;
// Actually propose the block.
let certificate = chain_client.retry_pending_block().await?;
// Now find the message ID in the certificate…
```

Reply:
Or we could avoid saving multiple times in some cases by extracting the code to a function. I.e. instead of

```rust
let result = publish();
save_wallet();
let bytecode_id = result?;
let result = create(bytecode_id);
save_wallet();
let application_id = result?
```

we'd do:

```rust
let result = publish_and_create();
save_wallet();
let application_id = result?;
```

Reply:
I see what you're saying but then we have to rename open_chain into prepare_open_chain etc. Alternatively, we could also create a wallet storage abstraction (or just callback) so that the client can "save" the wallet instead? Maybe something to rethink later